File: get-video/app/google_auth.py
import json
import os
from datetime import datetime, timezone

from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request

from .config import get_settings

SCOPES = [
    "https://www.googleapis.com/auth/gmail.readonly",
    "https://www.googleapis.com/auth/calendar.readonly",
]


def get_credentials() -> Credentials:
    settings = get_settings()
    tokens_path = settings.GOOGLE_TOKENS_PATH

    if not os.path.exists(tokens_path):
        raise FileNotFoundError(
            f"No Google tokens found at {tokens_path}. "
            "Run 'python auth_google.py' first to authenticate."
        )

    with open(tokens_path) as f:
        tokens = json.load(f)

    # Node.js googleapis stores expiry_date as epoch milliseconds
    expiry = None
    if "expiry_date" in tokens:
        expiry = datetime.fromtimestamp(tokens["expiry_date"] / 1000, tz=timezone.utc)
    elif "expiry" in tokens:
        # Python format: ISO string
        expiry = datetime.fromisoformat(tokens["expiry"])

    creds = Credentials(
        token=tokens.get("access_token"),
        refresh_token=tokens.get("refresh_token"),
        token_uri="https://oauth2.googleapis.com/token",
        client_id=settings.GOOGLE_CLIENT_ID,
        client_secret=settings.GOOGLE_CLIENT_SECRET,
        scopes=SCOPES,
        expiry=expiry,
    )

    # Refresh if expired
    if creds.expired and creds.refresh_token:
        creds.refresh(Request())
        _save_tokens(tokens_path, creds)

    return creds


def _save_tokens(tokens_path: str, creds: Credentials):
    with open(tokens_path) as f:
        existing = json.load(f)
    existing["access_token"] = creds.token
    if creds.expiry:
        existing["expiry_date"] = int(creds.expiry.timestamp() * 1000)
    with open(tokens_path, "w") as f:
        json.dump(existing, f, indent=2)

File: gmail/src/googleAuth.ts
import { google } from 'googleapis';
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';

const TOKENS_PATH = path.join(__dirname, '..', 'google_tokens.json');

const SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
  'https://www.googleapis.com/auth/calendar.readonly',
];

export function createOAuthClient() {
  return new google.auth.OAuth2(
    process.env.GOOGLE_CLIENT_ID!,
    process.env.GOOGLE_CLIENT_SECRET!,
    process.env.GOOGLE_REDIRECT_URI!
  );
}

export function getAuthUrl(): string {
  const client = createOAuthClient();
  return client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES,
    prompt: 'consent',
  });
}

export async function exchangeCodeForTokens(code: string): Promise<void> {
  const client = createOAuthClient();
  const { tokens } = await client.getToken(code);
  fs.writeFileSync(TOKENS_PATH, JSON.stringify(tokens, null, 2));
  console.log('✓ Google tokens saved to google_tokens.json');
}

export function getAuthenticatedClient() {
  if (!fs.existsSync(TOKENS_PATH)) {
    throw new Error(
      'No Google tokens found. Run "npm run auth:google" first to authenticate.'
    );
  }
  const tokens = JSON.parse(fs.readFileSync(TOKENS_PATH, 'utf-8'));
  const client = createOAuthClient();
  client.setCredentials(tokens);

  // Auto-refresh token when expired
  client.on('tokens', (newTokens) => {
    const existing = JSON.parse(fs.readFileSync(TOKENS_PATH, 'utf-8'));
    fs.writeFileSync(TOKENS_PATH, JSON.stringify({ ...existing, ...newTokens }, null, 2));
  });

  return client;
}

// Run directly: ts-node src/googleAuth.ts
if (require.main === module) {
  require('dotenv/config');
  runAuthFlow().catch(console.error);
}

export async function runAuthFlow(): Promise<void> {
  const url = getAuthUrl();
  console.log('\n--- Google OAuth Setup ---');
  console.log('1. Open this URL in your browser:\n');
  console.log(url);
  console.log('\n2. Log in and grant access.');
  console.log('3. You will be redirected to localhost (it will fail to load — that is normal).');
  console.log('4. Copy the "code" parameter from the URL.\n');

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const code = await new Promise<string>((resolve) => {
    rl.question('Paste the code here: ', (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });

  await exchangeCodeForTokens(code);
}
